libsolidity/analysis/TypeChecker.cpp

```diff
@@ -822,12 +822,12 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	{
 	case Type::Category::Array:
 		if (auto arrayType = dynamic_cast<ArrayType const*>(varType.get()))
-			if (
-				((arrayType->location() == DataLocation::Memory) ||
-				(arrayType->location() == DataLocation::CallData)) &&
-				!arrayType->validForCalldata()
-			)
-				m_errorReporter.typeError(_variable.location(), "Array is too large to be encoded.");
+		{
+			auto baseType = dynamic_cast<ArrayType const*>(arrayType->baseType().get());
+			if (arrayType->location() == DataLocation::Memory || arrayType->location() == DataLocation::CallData)
+				if ((baseType && !baseType->validForCalldata()) || !arrayType->validForCalldata())
```

Comment:
If the base type is not valid for calldata, why should the type itself be? This rather sounds like the change should be inside `validForCalldata`